From: SeongJae Park <sj@kernel.org>
Date: Sun, 1 Dec 2024 12:03:00 -0800
Subject: [PATCH] mm/damon/core: cancel damos_walk() if no scheme exists

If damos_walk() is called while there is not scheme, caller will be
blocked.  Check the case and cancel the walk.

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 mm/damon/core.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index 9b587c756044..e3a4614a0aee 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -2257,6 +2257,8 @@ static int kdamond_fn(void *data)
 		 */
 		if (!list_empty(&ctx->schemes))
 			kdamond_apply_schemes(ctx);
+		else
+			damos_walk_cancel(ctx);
 
 		sample_interval = ctx->attrs.sample_interval ?
 			ctx->attrs.sample_interval : 1;
-- 
2.39.5

